js/textInputModel.js

```diff
@@ -1,207 +1,192 @@
-define([
-  'core/js/models/questionModel'
-], function(QuestionModel) {
+import QuestionModel from 'core/js/models/questionModel';
+
+class TextInputModel extends QuestionModel {
+
+  init() {
+    super.init();
+
+    this.set('_genericAnswerIndexOffset', TextInputModel.genericAnswerIndexOffset);
+
+    this.setupQuestionItemIndexes();
+    this.checkCanSubmit();
+  }
+
+  setupQuestionItemIndexes() {
+    this.get('_items').forEach((item, index) => {
+      if (item._index === undefined) item._index = index;
+      if (item._answerIndex === undefined) item._answerIndex = -1;
+    });
+  }
+
+  restoreUserAnswers() {
+    if (!this.get('_isSubmitted')) return;
+
+    const userAnswer = this.get('_userAnswer');
+    const genericAnswers = this.get('_answers');
+    this.get('_items').forEach(item => {
+      const answerIndex = userAnswer[item._index];
+      if (answerIndex >= TextInputModel.genericAnswerIndexOffset) {
+        item.userAnswer = genericAnswers[answerIndex - TextInputModel.genericAnswerIndexOffset];
+        item._answerIndex = answerIndex;
+      } else if (answerIndex > -1) {
+        item.userAnswer = item._answers[answerIndex];
+        item._answerIndex = answerIndex;
+      } else {
+        if (item.userAnswer === undefined) item.userAnswer = '******';
+        item._answerIndex = -1;
+      }
+      if (item.userAnswer instanceof Array) item.userAnswer = item.userAnswer[0];
+    });
+
+    this.setQuestionAsSubmitted();
+    this.markQuestion();
+    this.setScore();
+    this.setupFeedback();
+  }
+
+  setupRandomisation() {
+    if (!this.get('_isRandom') || !this.get('_isEnabled')) return;
+
+    this.set('_items', _.shuffle(this.get('_items')));
+  }
+
+  // Use to check if the user is allowed to submit the question
+  canSubmit() {
+    // can submit if every item has user input
+    return this.get('_items').every(({ userAnswer }) => userAnswer);
+  }
+
+  setItemUserAnswer(itemIndex, userAnswer) {
+    const item = this.get('_items')[itemIndex];
+    item.userAnswer = userAnswer;
+    this.checkCanSubmit();
+  }
+
+  // Preserves the state for returning or showing the user's answers
+  storeUserAnswer() {
+    const items = this.get('_items');
+
+    this.isCorrect();
+
+    const userAnswer = new Array(items.length);
+    items.forEach(({ _index, _answerIndex }) => (userAnswer[_index] = _answerIndex));
+    this.set('_userAnswer', userAnswer);
+  }
+
+  isCorrect() {
+    if (this.get('_answers')) {
+      this.markGenericAnswers();
+    } else {
+      this.markSpecificAnswers();
+    }
+    // do we have any _isCorrect == false?
+    return this.get('_items').some(({ _isCorrect }) => !_isCorrect);
```

Comment:
Apologies, the confusing comment caused a typo; suggest removing.
```suggestion
    return this.get('_items').every(({ _isCorrect }) => _isCorrect);
```

Reply:
Please remove the comment.